File: plugins/openclaw/lib/scope-auto.ts
/**
 * Nexus Memory — Scope Auto-Inference (TypeScript port of scope_auto.py).
 *
 * Self-organizing memory areas (Nebo law 07.09: full automation or useless).
 * No user ever types a scope: centroids are computed from scoped canonical
 * points; a new memory inherits its area automatically on a CLEAR closest
 * match; a query that clearly belongs to one area unlocks that area.
 *
 * Conservative margins — under-tagging is harmless, over-tagging is what
 * we must avoid. Zero LLM cost: pure vector math. Fail-open everywhere:
 * any error or empty state = no areas = old behavior.
 */

import { fetchWithTimeout } from "./embedder.ts";

const CENTROID_TTL_MS = 60_000;
const MARGIN = 0.05;          // must beat runner-up by this cosine margin
const MIN_SIMILARITY = 0.72;  // absolute floor for the closest centroid

// Hard cap on scroll pages for the centroid fetch (mirrors MAX_SCROLL_PAGES in
// qdrant-client.ts). Prevents an "any-size collection" scan from unbounded
// fan-out while still sampling well beyond the first 1000 canonical points.
const MAX_SCROLL_PAGES = 5;

export type Centroids = Record<string, number[]>;

function dot(a: number[], b: number[]): number {
  if (a.length !== b.length) throw new Error("vector length mismatch");
  let s = 0;
  for (let i = 0; i < a.length; i++) s += a[i] * b[i];
  return s;
}

/**
 * L2-normalize into a NEW array (the input is left untouched); returns null
 * for a (near-)zero vector. Dimension mismatches are NOT handled here —
 * callers skip those before calling.
 */
function normalize(v: number[]): number[] | null {
  let mag = 0;
  for (const x of v) mag += x * x;
  mag = Math.sqrt(mag);
  if (mag < 1e-12) return null;
  return v.map((x) => x / mag);
}

/**
 * Fetch scoped canonical points and compute one centroid per scope.
 * Odd dimensions and near-zero vectors are skipped safely.
 */
export async function fetchCentroids(
  qdrantUrl: string,
  collection: string,
): Promise<Centroids> {
  // Paginated inline scroll (own fetch, NOT QdrantClient.scrollFiltered):
  // the centroid computation needs `with_vectors: true`, which the shared
  // helper does not request. Follows next_page_offset up to MAX_SCROLL_PAGES.
  const allPoints: Array<{
    vector?: number[] | null
    payload?: Record<string, unknown>
  }> = []
  let offset: unknown = undefined
  for (let page = 0; page < MAX_SCROLL_PAGES; page++) {
    const body: Record<string, unknown> = {
      filter: { must: [{ key: "lifecycle_status", match: { value: "canonical" } }] },
      with_payload: true,
      with_vectors: true,
      limit: 1000,
    }
    if (offset !== undefined && offset !== null) body.offset = offset
    const resp = await fetchWithTimeout(
      `${qdrantUrl}/collections/${collection}/points/scroll`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      },
    )
    if (!resp.ok) {
      throw new Error(`qdrant scroll failed: ${resp.status}`)
    }
    const data = (await resp.json()) as {
      result?: {
        points?: Array<{
          vector?: number[] | null
          payload?: Record<string, unknown>
        }>
        next_page_offset?: unknown
      }
    }
    for (const p of data.result?.points ?? []) allPoints.push(p)
    const next = data.result?.next_page_offset
    if (next === null || next === undefined) break
    offset = next
  }

  const sums: Record<string, { sum: number[]; n: number }> = {}
  let dim: number | null = null
  for (const p of allPoints) {
    const pl = p.payload ?? {}
    const scope = String(pl.scope ?? "default").trim().toLowerCase() || "default"
    if (scope === "default" || !Array.isArray(p.vector)) continue
    if (dim === null) dim = p.vector.length
    if (p.vector.length !== dim) continue // dimension mismatch → skip
    const norm = normalize(p.vector)
    if (!norm) continue
    const e = (sums[scope] ??= { sum: new Array(dim).fill(0), n: 0 })
    for (let i = 0; i < dim; i++) e.sum[i] += norm[i]
    e.n++
  }

  const cents: Centroids = {}
  for (const [scope, e] of Object.entries(sums)) {
    if (e.n === 0) continue
    const norm = normalize(e.sum)
    if (norm) cents[scope] = norm
  }
  return cents
}

/** Cached centroids with TTL; failures fail-open to {} (no areas). */
export class ScopeCentroidCache {
  private cache: Centroids = {}
  private at = 0
  private inflight: Promise<Centroids> | null = null
  // Generation counter: invalidate() bumps it so an already-running fetch
  // discards its (now stale) result instead of resurrecting the old cache.
  private gen = 0

  private qdrantUrl: string
  private collection: string

  constructor(qdrantUrl: string, collection: string) {
    this.qdrantUrl = qdrantUrl
    this.collection = collection
  }

  async get(): Promise<Centroids> {
    const now = Date.now()
    if (now - this.at < CENTROID_TTL_MS) return this.cache
    if (this.inflight) return this.inflight
    const gen = this.gen
    this.inflight = fetchCentroids(this.qdrantUrl, this.collection)
      .then((c) => {
        if (gen !== this.gen) return this.cache // invalidated mid-flight
        this.cache = c
        this.at = Date.now()
        return c
      })
      .catch((err) => {
        console.warn("scope_auto: centroid fetch failed — fail-open", err)
        if (gen === this.gen) this.at = Date.now() // don't hammer a down Qdrant
        return this.cache
      })
      .finally(() => {
        if (gen === this.gen) this.inflight = null
      })
    return this.inflight
  }

  invalidate(): void {
    this.at = 0
    this.gen++
    this.inflight = null
  }
}

/**
 * Closest centroid by cosine similarity.
 *
 * Centroids produced by `fetchCentroids` are ALREADY L2-normalized (each
 * source vector and then the summed centroid are normalized there), so the
 * dot product below IS the cosine similarity — no per-centroid re-normalization.
 */
function closestScope(
  vector: number[],
  cents: Centroids,
): { scope: string; margin: number; sim: number } | null {
  const norm = normalize(vector)
  if (!norm) return null
  let best: string | null = null
  let bestSim = -Infinity
  let second = -Infinity
  for (const [scope, c] of Object.entries(cents)) {
    const sim = dot(norm, c)
    if (sim > bestSim) {
      second = bestSim
      best = scope
      bestSim = sim
    } else if (sim > second) {
      second = sim
    }
  }
  if (!best) return null
  return { scope: best, margin: bestSim - second, sim: bestSim }
}

/** inferScope for WRITE path: clear closest match → scope; else 'default'. */
export function inferScope(vector: number[], cents: Centroids): string {
  const c = closestScope(vector, cents)
  if (!c) return "default"
  if (c.sim >= MIN_SIMILARITY && c.margin >= MARGIN) return c.scope
  return "default"
}

/**
 * prefetchFilterScopes for READ path: the query itself steers.
 * Returns null = no gating (old behavior); otherwise the allowed set
 * (always includes 'default' plus the clearly matched area).
 */
export function prefetchFilterScopes(
  vector: number[],
  cents: Centroids,
  manualScope: string,
): Set<string> | null {
  if (Object.keys(cents).length === 0) return null
  const c = closestScope(vector, cents)
  if (!c) return null
  if (c.sim >= MIN_SIMILARITY && c.margin >= MARGIN) {
    // Self-organizing: the query steers. A manual scope is ADDED to the
    // inferred area (not a replacement) — the agent keeps its home base
    // while still seeing the area the question belongs to. Parity with
    // Python prefetch_allowed_scopes().
    const allowed = new Set(["default", c.scope])
    if (manualScope) allowed.add(manualScope)
    return allowed
  }
  // No clear match → manual scope alone still gates (old behavior);
  // no manual scope → no gating at all (fail-open).
  if (manualScope) return new Set(["default", manualScope])
  return null // ambiguous → no gating
}

/**
 * Normalize a scope for the write path: trim, lowercase, then accept only
 * `[a-z0-9][a-z0-9-]{0,39}` (starts alphanumeric, at most 40 chars). Any
 * absent, empty or malformed input becomes "default" — fail-open, so an
 * invalid scope never blocks a write. Returns a NEW string; the input is
 * never modified.
 */
export function normalizeScope(scope: unknown): string {
  if (typeof scope === "string") {
    const s = scope.trim().toLowerCase()
    if (s && /^[a-z0-9][a-z0-9-]{0,39}$/.test(s)) return s
  }
  return "default"
}

/**
 * Config-scope validator: the same [a-z0-9-] contract as normalizeScope, but
 * FAIL-OPEN to "" for absent/invalid input. Callers use "" to mean "not
 * configured" (no gating) — distinct from normalizeScope's "default". This is
 * the single implementation shared by config.ts (cfg.scope + NEXUS_SCOPE).
 */
export function validateConfigScope(scope: unknown): string {
  if (typeof scope === "string") {
    const s = scope.trim().toLowerCase()
    if (s && /^[a-z0-9][a-z0-9-]{0,39}$/.test(s)) return s
  }
  return ""
}

File: plugins/claude-code/scripts/scope_auto.py
#!/usr/bin/env python3
"""Nexus Memory — Scope Auto-Inference (shared lib for Claude-Code hooks).

Self-organizing memory areas (Nebo law 07.09: full automation or useless).
No user ever types a scope: centroids are read from existing scoped canonical
points, and a memory/prompt is only tagged/filtered on a CLEAR match.

Conservative rule (identical thresholds to the MCP server's scope_auto.py):
  - cosine similarity must be >= SIM_THRESHOLD (default 0.65)
  - AND >= MARGIN (default 0.05) above the runner-up centroid
Otherwise: no area (default / no filtering). Fail-open everywhere.
Zero LLM cost: pure vector math.

Note: Claude-Code hooks are short-lived processes — centroids are fetched
per call (no cache). The fetch is one scroll request; fine at hook cadence.
"""

import logging
import os
import re
import time

QDRANT_URL = "http://localhost:6333"
COLLECTION = "nexus"

_SCOPE_RE = re.compile(r"^[a-z0-9][a-z0-9-]{0,39}$")

# Inference thresholds — read from the SAME env vars with the SAME defaults as
# src/nexus_memory/scope_auto.py (SCOPE_MATCH_THRESHOLD / SCOPE_MARGIN). H255:
# the hook used to hardcode 0.72 while the MCP server defaulted to 0.65, so
# the same memory could be tagged on one path and filtered out on the other.
SIM_THRESHOLD = float(os.getenv("NEXUS_SCOPE_AUTO_THRESHOLD", "0.65"))
MARGIN = float(os.getenv("NEXUS_SCOPE_AUTO_MARGIN", "0.05"))


def _cosine(a, b):
    """Cosine similarity; 0.0 on any degenerate input (fail-open)."""
    try:
        if not a or not b or len(a) != len(b):
            return 0.0
        num = sum(x * y for x, y in zip(a, b))
        na = sum(x * x for x in a) ** 0.5
        nb = sum(x * x for x in b) ** 0.5
        if na == 0 or nb == 0:
            return 0.0
        return num / (na * nb)
    except Exception:
        return 0.0


# Scroll pagination guard: hooks are short-lived, so cap the work hard.
_PAGE_LIMIT = 1000
_MAX_PAGES = 3


def fetch_centroids(qdrant_url: str = QDRANT_URL, collection: str = COLLECTION,
                    timeout: float = 3.0) -> dict:
    """Read scope centroids from Qdrant. Returns {scope: centroid_vector}.

    Fail-open: any error → {} (no areas → no filtering anywhere).
    Reads canonical scoped points only.

    Scope clause: `match: {"except": ["default"]}` is a POSITIVE match on
    existing non-default scope values (mirrors the server's _probe_scopes
    predicate). Without it, the ~8k default-only canonical points crowd the
    scoped points out of the scroll window → empty centroids → dead scope
    inference. Pagination follows next_page_offset for up to _MAX_PAGES.
    """
    import json
    import urllib.request

    try:
        scroll_filter = {"must": [
            {"key": "lifecycle_status", "match": {"value": "canonical"}},
            {"key": "scope", "match": {"except": ["default"]}},
        ]}
        points = []
        offset = None
        for _ in range(_MAX_PAGES):
            body = {
                "limit": _PAGE_LIMIT,
                "with_payload": True,
                "with_vector": True,
                "filter": scroll_filter,
            }
            if offset is not None:
                body["offset"] = offset
            req = urllib.request.Request(
                f"{qdrant_url}/collections/{collection}/points/scroll",
                data=json.dumps(body).encode(),
                headers={"Content-Type": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                data = json.loads(resp.read())
            result = data.get("result") or {}
            points.extend(result.get("points") or [])
            offset = result.get("next_page_offset")
            if not offset:
                break
    except Exception as exc:
        logging.info("scope_auto: centroid fetch failed (%s) — fail-open", exc)
        return {}

    # Parsing/aggregation stays INSIDE a guard: a malformed point (non-dict
    # payload, non-string scope, non-list/non-numeric vector) must degrade to
    # {} — never raise out of the short-lived hook.
    try:
        sums = {}
        counts = {}
        for p in points:
            if not isinstance(p, dict):
                continue
            payload = p.get("payload") or {}
            if not isinstance(payload, dict):
                continue
            raw_scope = payload.get("scope")
            scope = (raw_scope.strip().lower() if isinstance(raw_scope, str) else "") or "default"
            if scope == "default":
                continue
            vec = p.get("vector")
            if not isinstance(vec, list) or not vec:
                continue
            if not all(isinstance(x, (int, float)) and not isinstance(x, bool) for x in vec):
                continue
            if scope in sums:
                if len(vec) != len(sums[scope]):
                    continue  # dimension mismatch — skip safely
                sums[scope] = [x + y for x, y in zip(sums[scope], vec)]
                counts[scope] += 1
            else:
                sums[scope] = list(vec)
                counts[scope] = 1

        cents = {}
        for scope, total in sums.items():
            n = counts[scope]
            norm = sum(x * x for x in total) ** 0.5
            if n > 0 and norm > 0:
                cents[scope] = [x / norm for x in total]  # normalized centroid
        return cents
    except Exception as exc:
        logging.info("scope_auto: centroid parse failed (%s) — fail-open", exc)
        return {}


def infer_scope(vector, centroids: dict) -> str:
    """Clear-closest area for a vector, else 'default'. Pure math, fail-open."""
    try:
        if not vector or not centroids:
            return "default"
        scored = sorted(
            ((scope, _cosine(vector, c)) for scope, c in centroids.items()),
            key=lambda kv: kv[1], reverse=True,
        )
        top_scope, top = scored[0]
        runner_up = scored[1][1] if len(scored) > 1 else 0.0
        if top >= SIM_THRESHOLD and top - runner_up >= MARGIN:
            return top_scope
    except Exception as exc:
        logging.info("scope_auto: inference failed (%s) — fail-open", exc)
    return "default"


def prefetch_allowed_scopes(vector, centroids: dict, manual_scope: str):
    """Allowed scope set for auto-prefetch filtering.

    Returns a set (e.g. {'default', 'voice'}) when the query clearly belongs
    to one area, or None when there is NO clear match (→ no filtering,
    old behavior). A manual scope override always wins.

    H256: delegates to ``infer_scope`` instead of duplicating the scoring +
    threshold block (drift risk). Equivalence: the old code returned None
    exactly when no scope cleared BOTH the absolute threshold and the margin —
    which is precisely the condition under which ``infer_scope`` returns
    "default"; that is mapped back to None here.
    """
    try:
        inferred = infer_scope(vector, centroids)
        if inferred == "default":
            return None
        allowed = {"default", inferred}
        if manual_scope:
            allowed.add(manual_scope)
        return allowed
    except Exception as exc:
        logging.info("scope_auto: prefetch inference failed (%s) — fail-open", exc)
        return None